### PR TITLE
Virtual-Tour: Rotate before and after a new node, chain nodes

### PR DESCRIPTION
**Pull request by Mr-Kanister**
For my usecase I added the following features to the Virtual-Tour-Plugin:

- it is possible to overwrite the position to which the camera rotates to, when a link is clicked. This is useful when two links are too close to each other and thus the distance between them (to prevent them from overlapping) needs to be increased. To further ensure that the camera still animates to the right place, this option has been added.
- it is possible to specify a position to which the viewer rotates to, after the new node has been loaded. This is useful when you want to highlight specific things at a node in the tour.
- it is possible to automatically jump to another node after the first one has loaded. With this option it is possible to create "chains" of nodes, which is especially useful when a place is not of great importance in the tour, but it is helpful for orientation to include it as an intermediate step (e.g. on a stair landing, or in an anteroom).

I'd be very happy, if these would make it upstream. Please tell me, what you think about the implementation.

**Merge request checklist**

-   [x] All lints and tests pass.
-   [ ] If needed, the [documentation](https://github.com/mistic100/Photo-Sphere-Viewer/tree/main/docs) has been updated. (Not yet)

### CODE REVIEW

File: packages/virtual-tour-plugin/src/VirtualTourPlugin.ts [hunk 1]
```diff
@@ -342,6 +342,7 @@ export class VirtualTourPlugin extends AbstractConfigurablePlugin<
 
         const fromNode = this.state.currentNode;
         const fromLinkPosition = fromNode && fromLink ? this.__getLinkPosition(fromNode, fromLink) : null;
+        const rotateBeforeLoad = fromNode && fromLink && fromLink.rotateBeforeLoad ? fromLink.rotateBeforeLoad : fromLinkPosition;
```
Comment (mistic100) [inline packages/virtual-tour-plugin/src/VirtualTourPlugin.ts:345]:
```suggestion
        const rotateBeforeLoad = fromLink?.rotateBeforeLoad ?? fromLinkPosition;
```

File: packages/virtual-tour-plugin/src/VirtualTourPlugin.ts [hunk 2]
```diff
@@ -432,6 +433,24 @@ export class VirtualTourPlugin extends AbstractConfigurablePlugin<
 
                 return true;
             })
+            .then(() => {
+                const currentNode = this.state.currentNode;
+                const rotateAfterLoad = fromNode && fromLink && fromLink.rotateAfterLoad ? fromLink.rotateAfterLoad : null;
```
Comment (mistic100) [inline packages/virtual-tour-plugin/src/VirtualTourPlugin.ts:438]:
```suggestion
                const rotateAfterLoad = fromLink?.rotateAfterLoad;
```

File: packages/virtual-tour-plugin/src/VirtualTourPlugin.ts [hunk 3]
```diff
@@ -432,6 +433,24 @@ export class VirtualTourPlugin extends AbstractConfigurablePlugin<
 
                 return true;
             })
+            .then(() => {
```
Comment (mistic100) [inline packages/virtual-tour-plugin/src/VirtualTourPlugin.ts:436]: there is not point of chaining a promise here, the code can be put on line 433

File: packages/virtual-tour-plugin/src/VirtualTourPlugin.ts [hunk 4]
```diff
@@ -432,6 +433,24 @@ export class VirtualTourPlugin extends AbstractConfigurablePlugin<
...
+                const rotateAfterLoad = fromNode && fromLink && fromLink.rotateAfterLoad ? fromLink.rotateAfterLoad : null;
+                
+                Promise.resolve(rotateAfterLoad ? this.config.rotateSpeed : false)
+                    .then((speed) => {
+                        if (speed) {
+                            return this.viewer.animate({ ...rotateAfterLoad, speed });
+                        }
+                    })
+                    .then(() => {
+                        const nextNodeId = fromNode && fromLink ? fromLink.nextNodeId : null;
```
Comment (mistic100) [inline packages/virtual-tour-plugin/src/VirtualTourPlugin.ts:447]:
```suggestion
                        const nextNodeId = fromLink?.nextNodeId;
```

**Review by mistic100 — CHANGES_REQUESTED** [review]
- `rotateBeforeLoad` only works with `position`, when using `gps` the user won't know what is the final computed position. It would be better to have an `offsetPosition` which will be added to the link position (hence moving the marker/arrow but not where the viewer is rotated before going to the next node)

- `rotateAfterLoad` and `nextNodeId` are advanced features that can be easily implemented with event listeners, it is be much more versatile

**Comment by Mr-Kanister** [comment]
> `rotateAfterLoad` and `nextNodeId` are advanced features that can be easily implemented with event listeners, it is be much more versatile

Do you mean those features shouldn't be implemented, or should they be implemented with a NodeChanged event listener?

**Comment by mistic100** [comment]
> Do you mean those features shouldn't be implemented, or should they be implemented with a NodeChanged event listener?

They shouldn't be implemented

**Comment by Mr-Kanister** [comment]
I can't really understand why the build fails, I implemented your requests.

**Comment by mistic100** [comment]
don't worry that's on me, it is broken on the main branch too

**Pull request closed without merging**